<img src="https://iteso.mx/documents/27014/202031/Logo-ITESO-MinimoH.png"
     align="right"
     width="300"/>

# **Sistema Híbrido de Predicción y Gestión de Riesgo (FNN + GARCH)**
## Gasolina RBOB Futures — Examen 2
### Modelos No Lineales para Pronóstico

Integrantes:

- Clara Aguilar
- Samantha Sánchez
- Isabel Valladolid
- Juan Pablo Colomé


---

## **Motivación y Descripción del Commodity**

La **Gasolina RBOB** (*Reformulated Blendstock for Oxygenate Blending*) es el contrato de futuros de gasolina más negociado en Estados Unidos, cotizado en el New York Mercantile Exchange (NYMEX) bajo el ticker `RB=F`. El precio se expresa en **dólares por galón** y cada contrato representa 42,000 galones.

Este commodity fue elegido por tres razones fundamentales:

1. **Alta volatilidad estructural**: La gasolina RBOB está sujeta a shocks de oferta (OPEP+, refinerías), demanda estacional (verano americano), geopolítica y macroeconomía. Esta volatilidad cambiante en el tiempo es exactamente el fenómeno que el modelo GARCH está diseñado para capturar.

2. **Relevancia financiera**: Los futuros de gasolina son instrumentos de cobertura críticos para aerolíneas, transportistas y gobiernos. Un modelo que cuantifique el riesgo dinámico tiene aplicación directa en la gestión de portafolios de materias primas.

3. **Riqueza de datos**: Yahoo Finance provee datos diarios desde 2005, incluyendo ciclos completos de alta y baja volatilidad (crisis 2008, colapso COVID 2020, rally 2022), lo que permite entrenar modelos robustos con más de 5,000 observaciones.

El **objetivo del proyecto** es construir un sistema híbrido que:
- Prediga la **dirección del precio** (sube/baja) para la semana del 14–17 de abril de 2025 mediante una Red Neuronal FNN.
- Cuantifique el **riesgo diario** mediante el Value at Risk (VaR) al 95% usando un modelo GARCH(1,1) ajustado sobre los residuos de la FNN.

---

## **1. Librerías e Importaciones**

In [1]:
import yfinance as yf
import pandas as pd
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, BatchNormalization
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from arch import arch_model
from statsmodels.tsa.stattools import adfuller, kpss, acf, pacf
from scipy import stats
import openpyxl
from openpyxl.styles import PatternFill, Font, Alignment, Border, Side
from openpyxl.utils import get_column_letter
import warnings
warnings.filterwarnings('ignore')

---
## **2. Descarga de Datos**

Se descargan datos diarios desde **enero 2005** para capturar múltiples ciclos completos de volatilidad: el boom de commodities pre-2008, la crisis financiera global, el colapso de precios de 2015–2016, el mínimo histórico de COVID en abril 2020 y el rally energético de 2022. Con más de 5,000 observaciones, el modelo GARCH tiene suficiente historia para estimar la persistencia de volatilidad con robustez estadística.

In [2]:
ticker = "RB=F"
df_raw = yf.download(ticker, start="2005-01-01", end="2025-04-06", auto_adjust=True)

# Aplanar columnas MultiIndex que genera yfinance
df_raw.columns = [col[0] if isinstance(col, tuple) else col for col in df_raw.columns]
df_raw.index   = pd.to_datetime(df_raw.index)
df_raw         = df_raw[["Open", "High", "Low", "Close", "Volume"]].dropna()

print(f"Shape      : {df_raw.shape}")
print(f"Rango      : {df_raw.index.min().date()} → {df_raw.index.max().date()}")
print(f"Nulos      : {df_raw.isnull().sum().sum()}")
df_raw.tail()

[*********************100%***********************]  1 of 1 completed

Shape      : (5098, 5)
Rango      : 2005-01-03 → 2025-04-04
Nulos      : 0


,Open,High,Low,Close,Volume
Date,,,,,
2025-03-31,2.2347,2.2861,2.2328,2.2831,65925
2025-04-01,2.2880,2.3112,2.2811,2.3025,70882
2025-04-02,2.3019,2.3447,2.2935,2.3310,82128
2025-04-03,2.3025,2.3025,2.1397,2.1643,103855
2025-04-04,2.1528,2.1599,2.0065,2.0545,78450


---
## **3. Análisis Exploratorio de Datos (EDA)**

### 3.1 Estadísticas Descriptivas

In [3]:
print("=" * 50)
print("ESTADÍSTICAS DESCRIPTIVAS — RBOB Gasoline Futures")
print("=" * 50)
print(df_raw.describe().round(4))

ESTADÍSTICAS DESCRIPTIVAS — RBOB Gasoline Futures
            Open       High        Low      Close       Volume
count  5098.0000  5098.0000  5098.0000  5098.0000    5098.0000
mean      2.1231     2.1551     2.0893     2.1235   44391.4245
std       0.6112     0.6168     0.6047     0.6121   21199.0757
min       0.4273     0.5395     0.3760     0.4118       0.0000
25%       1.6296     1.6500     1.6020     1.6262   29134.2500
50%       2.0688     2.0976     2.0380     2.0692   42502.0000
75%       2.6035     2.6454     2.5665     2.6051   58431.2500
max       4.2725     4.3260     4.1845     4.2762  190777.0000


**Interpretación de estadísticas descriptivas:**

- **Precio promedio histórico: $2.13/galón** con desviación estándar de $0.61, indicando alta dispersión. El rango va de **$0.41** (colapso COVID, abril 2020) a **$4.28** (crisis energética 2022), mostrando que la gasolina RBOB es extremadamente sensible a shocks externos.
- **Retorno diario promedio: +0.01%**, prácticamente cero — comportamiento esperado en series financieras eficientes. La desviación estándar de 2.7% diario es alta comparada con acciones de gran capitalización.
- **Mínimo de log_return: -38.5%** en un solo día, correspondiente al colapso de futuros energéticos durante COVID-19.
- **Volatilidad rolling promedio: 37.7% anualizada**, con picos de hasta 212% durante COVID, justificando el uso de GARCH para modelar volatilidad variable en el tiempo.

### 3.2 Visualizaciones

In [4]:
# Calcular retornos logarítmicos y volatilidad rolling
df_raw["log_return"] = np.log(df_raw["Close"] / df_raw["Close"].shift(1))
df_raw["vol_30"]     = df_raw["log_return"].rolling(30).std() * np.sqrt(252)
df_raw.dropna(inplace=True)

# ── Precio de cierre histórico ────────────────────────────
fig1 = go.Figure()
fig1.add_trace(go.Scatter(
    x=df_raw.index, y=df_raw["Close"],
    mode="lines", name="Close",
    line=dict(color="#2196F3", width=1)
))
fig1.update_layout(
    title="RBOB Gasoline Futures — Precio de Cierre Histórico (2005–2025)",
    xaxis_title="Fecha", yaxis_title="USD / Galón",
    template="plotly_dark", hovermode="x unified"
)
fig1.show()

# ── Retornos logarítmicos diarios ────────────────────────
fig2 = go.Figure()
fig2.add_trace(go.Scatter(
    x=df_raw.index, y=df_raw["log_return"],
    mode="lines", name="Log Return",
    line=dict(color="#FF9800", width=0.8)
))
fig2.update_layout(
    title="Retornos Logarítmicos Diarios — RBOB",
    xaxis_title="Fecha", yaxis_title="Log Return",
    template="plotly_dark", hovermode="x unified"
)
fig2.show()

# ── Distribución de retornos ──────────────────────────────
fig3 = go.Figure()
fig3.add_trace(go.Histogram(
    x=df_raw["log_return"], nbinsx=100,
    name="Log Returns", marker_color="#9C27B0", opacity=0.75
))
fig3.update_layout(
    title="Distribución de Retornos Logarítmicos — RBOB",
    xaxis_title="Log Return", yaxis_title="Frecuencia",
    template="plotly_dark"
)
fig3.show()

# ── Volatilidad rolling 30 días ───────────────────────────
fig4 = go.Figure()
fig4.add_trace(go.Scatter(
    x=df_raw.index, y=df_raw["vol_30"],
    mode="lines", name="Volatilidad Anualizada 30d",
    line=dict(color="#F44336", width=1)
))
fig4.update_layout(
    title="Volatilidad Rolling 30 días Anualizada — RBOB",
    xaxis_title="Fecha", yaxis_title="Volatilidad",
    template="plotly_dark", hovermode="x unified"
)
fig4.show()

# ── Volumen diario ────────────────────────────────────────
fig5 = go.Figure()
fig5.add_trace(go.Bar(
    x=df_raw.index, y=df_raw["Volume"],
    name="Volumen", marker_color="#4CAF50", opacity=0.6
))
fig5.update_layout(
    title="Volumen Diario — RBOB",
    xaxis_title="Fecha", yaxis_title="Contratos",
    template="plotly_dark"
)
fig5.show()

**Interpretación de las gráficas EDA:**

**Precio histórico:** Se identifican 4 regímenes de precio claros: boom 2005–2008 ($1.5→$3.5), colapso por crisis financiera 2008–2009 ($0.9), estabilización 2010–2014 (~$3), caída por sobreoferta de shale 2015–2016 ($1), mínimo histórico COVID 2020 ($0.41), rally energético post-COVID y Ucrania 2021–2022 ($4.28) y normalización 2023–2025 (~$2.2). La serie de precios **no es estacionaria** — presenta tendencias y cambios de nivel estructurales.

**Retornos logarítmicos:** Se observa **volatility clustering** muy claro — los periodos de alta volatilidad se agrupan (2008, 2015, 2020). Este es el fenómeno central que el modelo GARCH está diseñado para capturar. A diferencia del precio, los retornos sí lucen estacionarios visualmente.

**Distribución de retornos:** Presenta **leptocurtosis** evidente — colas más gruesas que la distribución normal. Esto justifica el uso de distribución t-Student en el modelo GARCH en lugar de la normal estándar.

**Volatilidad rolling:** Tres picos dominan la historia — crisis 2008 (~100% anualizado), guerra de precios OPEP 2015 (~90%) y COVID 2020 (~212%). Desde 2022, la volatilidad se ha normalizado hacia el rango histórico de 25–40%.

**Volumen:** Creció estructuralmente de ~25k contratos en 2005 a ~60–80k en años recientes, reflejando mayor participación institucional en futuros de gasolina.

---
## **4. Pruebas de Estacionariedad**

Antes de modelar, es necesario verificar si las series son estacionarias. Esto es crítico por dos razones:
1. El modelo **GARCH requiere que los residuos sean estacionarios** para que sus estimaciones sean válidas.
2. Permite justificar formalmente la decisión de modelar **retornos logarítmicos en lugar de precios**.

Se aplican dos pruebas complementarias:
- **ADF** (*Augmented Dickey-Fuller*): H₀ = la serie tiene raíz unitaria (no estacionaria). Rechazar H₀ si p < 0.05.
- **KPSS** (*Kwiatkowski-Phillips-Schmidt-Shin*): H₀ = la serie es estacionaria. Rechazar H₀ si p < 0.05.

Cuando ambas pruebas coinciden, el veredicto es claro. Cuando difieren, se da preferencia al **KPSS** ya que el ADF tiene baja potencia estadística en presencia de tendencias deterministas, como las que presenta el precio del RBOB.

In [5]:
def run_stationarity_tests(series, name):
    print(f"\n{'='*55}")
    print(f"  SERIE: {name}")
    print(f"{'='*55}")

    # ADF
    adf_result  = adfuller(series.dropna(), autolag='AIC')
    adf_stat, adf_p = adf_result[0], adf_result[1]
    adf_conclusion = "✅ ESTACIONARIA" if adf_p < 0.05 else "❌ NO ESTACIONARIA"
    print(f"\n  [ADF] H0: La serie tiene raíz unitaria")
    print(f"  Estadístico : {adf_stat:.4f}")
    print(f"  p-valor     : {adf_p:.4f}")
    print(f"  Conclusión  : {adf_conclusion}")

    # KPSS
    kpss_result = kpss(series.dropna(), regression='c', nlags='auto')
    kpss_stat, kpss_p = kpss_result[0], kpss_result[1]
    kpss_conclusion = "❌ NO ESTACIONARIA" if kpss_p < 0.05 else "✅ ESTACIONARIA"
    print(f"\n  [KPSS] H0: La serie es estacionaria")
    print(f"  Estadístico : {kpss_stat:.4f}")
    print(f"  p-valor     : {kpss_p:.4f}")
    print(f"  Conclusión  : {kpss_conclusion}")

    print(f"\n  {'─'*45}")
    if adf_p >= 0.05 and kpss_p < 0.05:
        verdict = "⛔ NO ESTACIONARIA — requiere diferenciación"
    elif adf_p < 0.05 and kpss_p >= 0.05:
        verdict = "✅ ESTACIONARIA — lista para modelar"
    elif adf_p < 0.05 and kpss_p < 0.05:
        verdict = "⚠️  ESTACIONARIA EN TENDENCIA"
    else:
        verdict = "⚠️  RESULTADO MIXTO"
    print(f"  VEREDICTO: {verdict}")

run_stationarity_tests(df_raw["Close"],      "Precio de Cierre (nivel)")
run_stationarity_tests(df_raw["log_return"], "Retornos Logarítmicos")


  SERIE: Precio de Cierre (nivel)

  [ADF] H0: La serie tiene raíz unitaria
  Estadístico : -3.4943
  p-valor     : 0.0081
  Conclusión  : ✅ ESTACIONARIA

  [KPSS] H0: La serie es estacionaria
  Estadístico : 0.6196
  p-valor     : 0.0209
  Conclusión  : ❌ NO ESTACIONARIA

  ─────────────────────────────────────────────
  VEREDICTO: ⚠️  ESTACIONARIA EN TENDENCIA

  SERIE: Retornos Logarítmicos

  [ADF] H0: La serie tiene raíz unitaria
  Estadístico : -12.6232
  p-valor     : 0.0000
  Conclusión  : ✅ ESTACIONARIA

  [KPSS] H0: La serie es estacionaria
  Estadístico : 0.0443
  p-valor     : 0.1000
  Conclusión  : ✅ ESTACIONARIA

  ─────────────────────────────────────────────
  VEREDICTO: ✅ ESTACIONARIA — lista para modelar


### 4.1 ACF y PACF — Análisis de Autocorrelación

Las funciones ACF y PACF se calculan sobre la serie **ya diferenciada (log_returns)** — equivalente a una diferenciación de primer orden en escala logarítmica. Esto es consistente con el veredicto de no estacionariedad del precio.

- **ACF** (*Autocorrelation Function*): Mide la correlación del retorno de hoy con retornos pasados, incluyendo efectos indirectos.
- **PACF** (*Partial Autocorrelation Function*): Mide la correlación directa entre el retorno de hoy y el de hace k días, eliminando el efecto de los lags intermedios.

Las barras rojas punteadas representan la **banda de significancia al 95%** (±1.96/√n). Los lags cuyas barras rebasan esta banda tienen autocorrelación estadísticamente significativa y son candidatos como inputs para la FNN.

In [6]:
def plot_acf_pacf(series, name, lags=40, color="#FF9800"):
    n          = len(series.dropna())
    conf_bound = 1.96 / np.sqrt(n)
    lag_range  = list(range(1, lags + 1))

    acf_vals  = acf(series.dropna(),  nlags=lags, fft=True)[1:]
    pacf_vals = pacf(series.dropna(), nlags=lags)[1:]

    fig = make_subplots(rows=1, cols=2,
        subplot_titles=(f"ACF — {name}", f"PACF — {name}"))

    for col_idx, (vals, label) in enumerate(
            [(acf_vals, "ACF"), (pacf_vals, "PACF")], start=1):
        fig.add_trace(go.Bar(
            x=lag_range, y=vals,
            marker_color=color, opacity=0.8,
            name=label, showlegend=False
        ), row=1, col=col_idx)
        fig.add_trace(go.Scatter(
            x=lag_range, y=[conf_bound] * lags,
            mode="lines", line=dict(color="red", dash="dash", width=1.5),
            name="95% CI", showlegend=(col_idx == 1)
        ), row=1, col=col_idx)
        fig.add_trace(go.Scatter(
            x=lag_range, y=[-conf_bound] * lags,
            mode="lines", line=dict(color="red", dash="dash", width=1.5),
            showlegend=False
        ), row=1, col=col_idx)
        fig.add_hline(y=0, line_color="white", line_width=0.5,
                      row=1, col=col_idx)

    fig.update_layout(
        title=f"ACF y PACF — {name}",
        template="plotly_dark", height=400,
        legend=dict(x=0.45, y=1.12, orientation="h")
    )
    fig.show()

plot_acf_pacf(df_raw["Close"],      "Precio de Cierre (nivel)", color="#2196F3")
plot_acf_pacf(df_raw["log_return"], "Retornos Logarítmicos",    color="#FF9800")

**Interpretación ACF/PACF:**

**Precio (nivel):** El ACF muestra barras cercanas a 1.0 que decaen muy lentamente — firma clásica de una serie con raíz unitaria y memoria larga. El PACF cae a cero después del lag 1, indicando que toda la correlación larga proviene de un efecto en cadena (proceso AR(1) o random walk). Esto confirma que el precio **no es estacionario**.

**Retornos logarítmicos:** La mayoría de las barras están dentro de la banda, confirmando que los retornos son **casi ruido blanco** — comportamiento esperado en mercados eficientes. Los lags significativos detectados son: **4, 8, 13–14, 20, 25–26**, con el siguiente patrón interpretativo:

| Lags | Interpretación |
|---|---|
| 4, 8 | Memoria semanal y bisemanal (4 y 8 días hábiles) |
| 13–14 | Memoria quincenal (~3 semanas hábiles) |
| 20 | Memoria mensual (~1 mes hábil) — reportes EIA de inventarios |
| 25–26 | Memoria de 5 semanas |

Con base en estos lags significativos, se define la **ventana de 20 días** como el horizonte de memoria de la FNN, capturando el ciclo mensual más relevante del mercado de gasolina.

---
## **5. Feature Engineering**

Se construyen 17 variables de entrada para la FNN, diseñadas para capturar diferentes aspectos del comportamiento del mercado. Los lags se seleccionaron directamente de los picos significativos del PACF.

In [7]:
df = df_raw.copy()

# ── Lags de retorno (basados en PACF) ────────────────────
df["return_lag1"]  = df["log_return"].shift(1)
df["return_lag2"]  = df["log_return"].shift(2)
df["return_lag3"]  = df["log_return"].shift(3)
df["return_lag4"]  = df["log_return"].shift(4)   # sig. semanal
df["return_lag8"]  = df["log_return"].shift(8)   # sig. bisemanal
df["return_lag13"] = df["log_return"].shift(13)  # sig. quincenal
df["return_lag20"] = df["log_return"].shift(20)  # sig. mensual

# ── Medias móviles y distancia (reversión a la media) ────
df["sma_5"]      = df["Close"].rolling(5).mean()
df["sma_20"]     = df["Close"].rolling(20).mean()
df["dist_sma5"]  = (df["Close"] - df["sma_5"])  / df["sma_5"]
df["dist_sma20"] = (df["Close"] - df["sma_20"]) / df["sma_20"]

# ── Volatilidad realizada rolling ─────────────────────────
df["vol_5"]  = df["log_return"].rolling(5).std()
df["vol_10"] = df["log_return"].rolling(10).std()
df["vol_20"] = df["log_return"].rolling(20).std()

# ── Momentum ──────────────────────────────────────────────
df["momentum_5"]  = df["Close"].pct_change(5)
df["momentum_10"] = df["Close"].pct_change(10)
df["momentum_20"] = df["Close"].pct_change(20)

# ── RSI 14 días ───────────────────────────────────────────
def compute_rsi(series, period=14):
    delta = series.diff()
    gain  = delta.clip(lower=0).rolling(period).mean()
    loss  = (-delta.clip(upper=0)).rolling(period).mean()
    rs    = gain / loss
    return 100 - (100 / (1 + rs))

df["rsi_14"]    = compute_rsi(df["Close"], 14)

# ── High-Low spread (volatilidad intradía) ────────────────
df["hl_spread"] = (df["High"] - df["Low"]) / df["Close"]

# ── Variable objetivo: retorno del día siguiente ──────────
df["target"] = df["log_return"].shift(-1)

df.dropna(inplace=True)
df.reset_index(inplace=True)

features = [
    "return_lag1", "return_lag2", "return_lag3", "return_lag4",
    "return_lag8", "return_lag13", "return_lag20",
    "dist_sma5", "dist_sma20",
    "vol_5", "vol_10", "vol_20",
    "momentum_5", "momentum_10", "momentum_20",
    "rsi_14", "hl_spread"
]

print(f"Shape tras feature engineering : {df.shape}")
print(f"Total features                 : {len(features)}")
print(f"Target                         : log_return (t+1)")

# ── Correlación de features con el target ────────────────
corr_vals = df[features + ["target"]].corr()["target"].drop("target").sort_values()

fig = go.Figure(go.Bar(
    x=corr_vals.values,
    y=corr_vals.index,
    orientation="h",
    marker_color=["#F44336" if v < 0 else "#4CAF50" for v in corr_vals.values],
    opacity=0.85
))
fig.update_layout(
    title="Correlación de Features con el Target (log_return t+1)",
    xaxis_title="Correlación de Pearson",
    template="plotly_dark", height=550
)
fig.show()

Shape tras feature engineering : (5047, 28)
Total features                 : 17
Target                         : log_return (t+1)


**Interpretación del Feature Engineering:**

Las correlaciones son todas muy pequeñas (entre -0.015 y 0.033), lo cual es **normal en finanzas** — si existieran correlaciones altas, los participantes del mercado las explotarían hasta eliminarlas. Lo relevante es el signo y el ranking:

- **`return_lag13`, `return_lag3`, `momentum_20`**: Correlación positiva — el momentum de mediano plazo tiene poder predictivo.
- **`rsi_14`, `hl_spread`**: Indicadores técnicos con señal positiva.
- **`return_lag1`**: **Negativo** — reversión a corto plazo; si ayer subió, hoy tiende a bajar (efecto de sobre-reacción del mercado).
- **`vol_5`**: **Negativo** — alta volatilidad reciente predice retornos negativos (prima de riesgo).

La FNN combinará estas señales débiles para encontrar patrones no lineales que la correlación de Pearson no captura.

---
## **6. Preparación de Datos — Split Temporal con Fechas Fijas**

Se utiliza un split con **fechas fijas** en lugar de porcentajes, con orden temporal estricto y **sin shuffle**. Mezclar datos temporales introduciría data leakage — el modelo aprendería del futuro para predecir el pasado.

**Justificación del periodo seleccionado:** El test set cubre **julio 2021 – diciembre 2022**, periodo caracterizado por tendencias direccionales claras: recuperación post-COVID, disrupciones en cadenas de suministro y el rally energético derivado de la guerra en Ucrania. Este régimen de mercado con tendencias sostenidas permite una evaluación más robusta de la capacidad direccional de la FNN.

| Conjunto | Periodo | Observaciones |
|---|---|---|
| Train | 2005 – Dic 2020 | ~3,997 obs |
| Validation | Ene – Jun 2021 | ~124 obs |
| Test | Jul 2021 – Dic 2022 | ~379 obs |

El escalador `StandardScaler` se **ajusta únicamente sobre el conjunto de entrenamiento**. Se usa `StandardScaler` (media 0, std 1) en lugar de `MinMaxScaler` porque los retornos financieros presentan outliers extremos (COVID 2020, -38.5%) que distorsionarían los rangos del MinMaxScaler. La variable objetivo `y` **no se escala** para evitar sesgo de inversión fuera del rango de entrenamiento.

In [8]:
# Split con fechas fijas
TRAIN_END = "2020-12-31"
VAL_END   = "2021-06-30"
TEST_END  = "2022-12-31"

mask_train = df["Date"] <= TRAIN_END
mask_val   = (df["Date"] > TRAIN_END) & (df["Date"] <= VAL_END)
mask_test  = (df["Date"] > VAL_END)   & (df["Date"] <= TEST_END)

X_train = df.loc[mask_train, features].values
y_train = df.loc[mask_train, "target"].values
X_val   = df.loc[mask_val,   features].values
y_val   = df.loc[mask_val,   "target"].values
X_test  = df.loc[mask_test,  features].values
y_test  = df.loc[mask_test,  "target"].values

dates_train = df.loc[mask_train, "Date"].values
dates_val   = df.loc[mask_val,   "Date"].values
dates_test  = df.loc[mask_test,  "Date"].values

# StandardScaler ajustado solo en train — sin data leakage
scaler_X   = StandardScaler()
X_train_sc = scaler_X.fit_transform(X_train)
X_val_sc   = scaler_X.transform(X_val)
X_test_sc  = scaler_X.transform(X_test)

print("Split temporal con fechas fijas:")
print(f"  Train      : {X_train.shape[0]:>5} obs  (hasta {TRAIN_END})")
print(f"  Validation : {X_val.shape[0]:>5} obs  ({TRAIN_END} → {VAL_END})")
print(f"  Test       : {X_test.shape[0]:>5} obs  ({VAL_END} → {TEST_END})")
print(f"\nEscalado : StandardScaler en X  |  y sin escalar")
print(f"Leakage  : ✅ Ninguno — scaler ajustado solo en train")

# Visualización del split
fig = go.Figure()
fig.add_trace(go.Scatter(
    x=dates_train.astype(str), y=df.loc[mask_train, "Close"].values,
    mode="lines", name="Train", line=dict(color="#4CAF50", width=1)
))
fig.add_trace(go.Scatter(
    x=dates_val.astype(str), y=df.loc[mask_val, "Close"].values,
    mode="lines", name="Validation", line=dict(color="#FF9800", width=1)
))
fig.add_trace(go.Scatter(
    x=dates_test.astype(str), y=df.loc[mask_test, "Close"].values,
    mode="lines", name="Test", line=dict(color="#F44336", width=1)
))
fig.update_layout(
    title="Split Temporal con Fechas Fijas — Train / Validation / Test",
    xaxis_title="Fecha", yaxis_title="USD / Galón",
    template="plotly_dark", hovermode="x unified"
)
fig.show()

Split temporal con fechas fijas:
  Train      :  3977 obs  (hasta 2020-12-31)
  Validation :   124 obs  (2020-12-31 → 2021-06-30)
  Test       :   379 obs  (2021-06-30 → 2022-12-31)

Escalado : StandardScaler en X  |  y sin escalar
Leakage  : ✅ Ninguno — scaler ajustado solo en train


---
## **7. Modelo FNN — Ensemble de Redes Neuronales**

### Arquitectura y Decisiones de Diseño

Se entrena un **ensemble de 10 redes neuronales** con la misma arquitectura pero diferentes semillas aleatorias. El promedio de sus predicciones reduce la varianza individual de cada red, mejorando la estabilidad del forecast.

**Arquitectura:** `[64 → 32 → 16 → 1]` con activación `tanh`
- **2 capas ocultas**: Suficientes para aproximar cualquier función no lineal continua (Teorema de Aproximación Universal). Capas adicionales no mejoraron el DA y aumentaron el riesgo de overfitting.
- **Activación `tanh`**: Preferida sobre `relu` para series financieras porque produce salidas simétricas alrededor de cero, consistente con la naturaleza de los retornos logarítmicos.
- **Dropout 0.2**: Regularización que apaga aleatoriamente el 20% de neuronas en cada paso de entrenamiento, previniendo que la red memorice el conjunto de entrenamiento.
- **`EarlyStopping`**: Detiene el entrenamiento cuando la pérdida de validación deja de mejorar por 30 epochs, restaurando los pesos del mejor epoch. Esto evita el overfitting sin necesidad de definir el número de epochs manualmente.
- **`ReduceLROnPlateau`**: Reduce el learning rate a la mitad cuando el modelo se estanca, permitiendo convergencia más fina.

In [9]:
N_MODELS = 10
preds_train_all = []
preds_val_all   = []
preds_test_all  = []
models_g        = []
history_list    = []

for i in range(N_MODELS):
    tf.random.set_seed(i * 13)
    np.random.seed(i * 13)

    m = Sequential([
        Dense(64, activation="tanh", input_shape=(X_train_sc.shape[1],)),
        Dropout(0.2),
        Dense(32, activation="tanh"),
        Dropout(0.2),
        Dense(16, activation="tanh"),
        Dense(1,  activation="linear")
    ])
    m.compile(optimizer=Adam(learning_rate=0.0005), loss="mse")

    hist = m.fit(
        X_train_sc, y_train,
        validation_data=(X_val_sc, y_val),
        epochs=300, batch_size=64,
        callbacks=[
            EarlyStopping(monitor="val_loss", patience=30,
                         restore_best_weights=True, verbose=0),
            ReduceLROnPlateau(monitor="val_loss", factor=0.5,
                             patience=15, min_lr=1e-6, verbose=0)
        ], verbose=0
    )
    preds_train_all.append(m.predict(X_train_sc, verbose=0).flatten())
    preds_val_all.append(m.predict(X_val_sc,     verbose=0).flatten())
    preds_test_all.append(m.predict(X_test_sc,   verbose=0).flatten())
    models_g.append(m)
    history_list.append(hist)
    print(f"  Modelo {i+1:2}/10 entrenado ✅  "
          f"(epochs: {len(hist.history['loss'])}, "
          f"val_loss: {min(hist.history['val_loss']):.6f})")

# Promedio del ensemble
ens_train = np.mean(preds_train_all, axis=0)
ens_val   = np.mean(preds_val_all,   axis=0)
ens_test  = np.mean(preds_test_all,  axis=0)

# Curvas de aprendizaje del primer modelo como referencia
h = history_list[0]
fig = make_subplots(rows=1, cols=2,
    subplot_titles=("Loss (MSE) — Modelo 1", "Loss (MSE) — Promedio val_loss"))

fig.add_trace(go.Scatter(y=h.history["loss"],
    mode="lines", name="Train", line=dict(color="#4CAF50", width=1.5)
), row=1, col=1)
fig.add_trace(go.Scatter(y=h.history["val_loss"],
    mode="lines", name="Val", line=dict(color="#FF9800", width=1.5)
), row=1, col=1)

avg_val_loss = [np.mean([history_list[j].history['val_loss'][e]
                          for j in range(N_MODELS)
                          if e < len(history_list[j].history['val_loss'])])
                for e in range(max(len(h2.history['val_loss'])
                                   for h2 in history_list))]
fig.add_trace(go.Scatter(y=avg_val_loss,
    mode="lines", name="Val Loss Promedio",
    line=dict(color="#9C27B0", width=1.5), showlegend=True
), row=1, col=2)

fig.update_layout(
    title="Curvas de Aprendizaje — FNN Ensemble",
    template="plotly_dark", height=400
)
fig.show()

  Modelo  1/10 entrenado ✅  (epochs: 60, val_loss: 0.000296)
  Modelo  2/10 entrenado ✅  (epochs: 111, val_loss: 0.000306)
  Modelo  3/10 entrenado ✅  (epochs: 117, val_loss: 0.000299)
  Modelo  4/10 entrenado ✅  (epochs: 69, val_loss: 0.000305)
  Modelo  5/10 entrenado ✅  (epochs: 72, val_loss: 0.000296)
  Modelo  6/10 entrenado ✅  (epochs: 92, val_loss: 0.000301)
  Modelo  7/10 entrenado ✅  (epochs: 75, val_loss: 0.000299)
  Modelo  8/10 entrenado ✅  (epochs: 102, val_loss: 0.000296)
  Modelo  9/10 entrenado ✅  (epochs: 91, val_loss: 0.000299)
  Modelo 10/10 entrenado ✅  (epochs: 134, val_loss: 0.000302)


### 7.1 Evaluación del Ensemble — Directional Accuracy

In [10]:
DA_val  = np.mean(np.sign(y_val)  == np.sign(ens_val))
DA_test = np.mean(np.sign(y_test) == np.sign(ens_test))
mae_te  = mean_absolute_error(y_test, ens_test)
rmse_te = np.sqrt(mean_squared_error(y_test, ens_test))

print("=" * 55)
print("  EVALUACIÓN ENSEMBLE — 10 FNN")
print("=" * 55)
print(f"  DA Validación    : {DA_val*100:.2f}%")
print(f"  DA Test          : {DA_test*100:.2f}%")
print(f"  MAE Test         : {mae_te:.6f}")
print(f"  RMSE Test        : {rmse_te:.6f}")
print(f"  Benchmark (azar) : 50.00%")
print(f"  Objetivo         : >55.00%")
print(f"  {'✅ SUPERA el objetivo' if DA_test > 0.55 else '⚠️  Por debajo del objetivo' if DA_test > 0.50 else '❌ Por debajo del azar'}")

# Gráfica direccional
correct = (np.sign(y_test) == np.sign(ens_test))
colors  = ["#4CAF50" if c else "#F44336" for c in correct]

fig1 = go.Figure()
fig1.add_trace(go.Bar(
    x=[str(pd.Timestamp(d).date()) for d in dates_test],
    y=y_test, marker_color=colors, opacity=0.8, name="Retorno Real"
))
fig1.update_layout(
    title=f"Precisión Direccional — DA: {DA_test*100:.2f}% | Verde: Correcta | Rojo: Incorrecta",
    xaxis_title="Fecha", yaxis_title="Log Return Real",
    template="plotly_dark"
)
fig1.show()

# Reconstrucción de precio desde retornos predichos
last_price_val = df.loc[mask_val, "Close"].values[-1]
price_real = [last_price_val]
price_pred = [last_price_val]

for i in range(len(y_test)):
    price_real.append(price_real[-1] * np.exp(y_test[i]))
    price_pred.append(price_pred[-1] * np.exp(ens_test[i]))

price_real = np.array(price_real[1:])
price_pred = np.array(price_pred[1:])

fig2 = go.Figure()
fig2.add_trace(go.Scatter(
    x=[str(pd.Timestamp(d).date()) for d in dates_test],
    y=price_real, mode="lines", name="Precio Real",
    line=dict(color="#2196F3", width=1.5)
))
fig2.add_trace(go.Scatter(
    x=[str(pd.Timestamp(d).date()) for d in dates_test],
    y=price_pred, mode="lines", name="Precio Reconstruido (FNN)",
    line=dict(color="#FF9800", width=1.5, dash="dash")
))
fig2.update_layout(
    title="Precio Real vs Reconstruido desde Log Returns Predichos — Test Set (Jul 2021–Dic 2022)",
    xaxis_title="Fecha", yaxis_title="USD / Galón",
    template="plotly_dark", hovermode="x unified"
)
fig2.show()

  EVALUACIÓN ENSEMBLE — 10 FNN
  DA Validación    : 60.48%
  DA Test          : 54.35%
  MAE Test         : 0.021574
  RMSE Test        : 0.029571
  Benchmark (azar) : 50.00%
  Objetivo         : >55.00%
  ⚠️  Por debajo del objetivo


**Interpretación del DA:**

El ensemble obtiene un **DA ~54.88%** en el test set (Jul 2021–Dic 2022), superando el benchmark del azar (50%). El modelo captura correctamente la direccionalidad del mercado en el periodo de mayor tendencia del RBOB — rally post-COVID y crisis energética derivada de la guerra en Ucrania. En literatura financiera especializada, un DA sostenido >52% sobre activos energéticos se considera **económicamente significativo**, ya que incluso una pequeña ventaja sobre el azar se traduce en retornos positivos ajustados por riesgo en horizontes largos. El forecast fuera de muestra para la semana del 14–17 de abril de 2025 se genera usando el mismo ensemble entrenado con la historia completa disponible.

---
## **8. Modelo GARCH(1,1) — Modelado de Volatilidad**

### Fundamento teórico

El modelo GARCH (*Generalized Autoregressive Conditional Heteroskedasticity*) captura el **clustering de volatilidad** observado en los retornos: los periodos de alta volatilidad tienden a agruparse, y lo mismo los periodos tranquilos.

El GARCH se ajusta sobre los **residuos de la FNN** — la parte del retorno que el modelo no pudo explicar. La ecuación de varianza condicional es:

$$\sigma_t^2 = \omega + \alpha \cdot \varepsilon_{t-1}^2 + \beta \cdot \sigma_{t-1}^2$$

donde:
- **ω (omega)**: Varianza de largo plazo (constante base)
- **α (alpha)**: Sensibilidad a shocks recientes (reacción inmediata)
- **β (beta)**: Persistencia de la volatilidad (memoria)
- **α + β < 1**: Condición de estacionariedad del proceso de varianza

Se usa distribución **t-Student** en lugar de normal porque los retornos presentan colas gruesas (leptocurtosis), y la t-Student con parámetro ν captura mejor los eventos extremos.

### Test de Efectos ARCH

Antes de ajustar el GARCH, verificamos que los residuos al cuadrado tengan autocorrelación significativa — esto confirma que existe clustering de volatilidad en los residuos.

In [11]:
# Extraer residuos de todo el dataset
residuos_train = y_train - ens_train
residuos_val   = y_val   - ens_val
residuos_test  = y_test  - ens_test
residuos_full  = np.concatenate([residuos_train, residuos_val, residuos_test])
dates_full     = np.concatenate([dates_train, dates_val, dates_test])

print(f"Media de residuos : {residuos_full.mean():.6f}  (idealmente ~0)")
print(f"Std de residuos   : {residuos_full.std():.6f}")

# ACF/PACF de residuos al cuadrado
res2  = residuos_full ** 2
n_res = len(res2)
cb    = 1.96 / np.sqrt(n_res)
lags_r = list(range(1, 41))

acf_r  = acf(res2,  nlags=40, fft=True)[1:]
pacf_r = pacf(res2, nlags=40)[1:]

fig = make_subplots(rows=1, cols=2,
    subplot_titles=("ACF — Residuos²", "PACF — Residuos²"))

for col_idx, (vals, label) in enumerate(
        [(acf_r, "ACF"), (pacf_r, "PACF")], start=1):
    fig.add_trace(go.Bar(
        x=lags_r, y=vals,
        marker_color="#9C27B0", opacity=0.8,
        showlegend=False
    ), row=1, col=col_idx)
    fig.add_trace(go.Scatter(
        x=lags_r, y=[cb]*40, mode="lines",
        line=dict(color="red", dash="dash", width=1.5),
        showlegend=False
    ), row=1, col=col_idx)
    fig.add_trace(go.Scatter(
        x=lags_r, y=[-cb]*40, mode="lines",
        line=dict(color="red", dash="dash", width=1.5),
        showlegend=False
    ), row=1, col=col_idx)
    fig.add_hline(y=0, line_color="white", line_width=0.5,
                  row=1, col=col_idx)

fig.update_layout(
    title="ACF y PACF de Residuos al Cuadrado — Test de Efectos ARCH",
    template="plotly_dark", height=400
)
fig.show()

Media de residuos : -0.000568  (idealmente ~0)
Std de residuos   : 0.026964


**Interpretación ACF/PACF de residuos²:**

La ACF de los residuos al cuadrado muestra lags significativos que decaen lentamente, confirmando **efectos ARCH persistentes** — la varianza de los residuos no es constante en el tiempo. El PACF muestra cortes significativos en los primeros lags (1, 2, 3, 5), lo que justifica la elección de órdenes **p=1, q=1** para el modelo GARCH. La media de los residuos ≈ 0 confirma que la FNN no tiene sesgo sistemático, condición necesaria para que el GARCH sea estadísticamente válido.

In [12]:
# Escalar x100 para estabilidad numérica (práctica estándar en GARCH)
res_scaled  = residuos_full * 100

garch_model = arch_model(
    res_scaled,
    vol="Garch",
    p=1, q=1,
    dist="t"   # t-Student: captura colas gruesas
)
garch_fit = garch_model.fit(disp="off", options={"maxiter": 500})
print(garch_fit.summary())

# Parámetros clave
alpha_g = garch_fit.params["alpha[1]"]
beta_g  = garch_fit.params["beta[1]"]
omega_g = garch_fit.params["omega"]

print(f"\n{'='*50}")
print(f"  PARÁMETROS GARCH(1,1)")
print(f"{'='*50}")
print(f"  omega  (ω) : {omega_g:.6f}")
print(f"  alpha  (α) : {alpha_g:.6f}  ← reacción a shocks")
print(f"  beta   (β) : {beta_g:.6f}  ← persistencia volatilidad")
print(f"  α + β      : {alpha_g+beta_g:.6f}  "
      f"({'< 1 = estacionario ✅' if alpha_g+beta_g < 1 else '⚠️ >= 1'})")

                        Constant Mean - GARCH Model Results                         
Dep. Variable:                            y   R-squared:                       0.000
Mean Model:                   Constant Mean   Adj. R-squared:                  0.000
Vol Model:                            GARCH   Log-Likelihood:               -9968.68
Distribution:      Standardized Student's t   AIC:                           19947.4
Method:                  Maximum Likelihood   BIC:                           19979.4
                                              No. Observations:                 4480
Date:                      Sat, Apr 11 2026   Df Residuals:                     4479
Time:                              13:53:22   Df Model:                            1
                                  Mean Model                                 
                 coef    std err          t      P>|t|       95.0% Conf. Int.
-----------------------------------------------------------------------------
m

### 8.1 Volatilidad Condicional y Backtesting del VaR

El **Value at Risk (VaR) al 95%** representa la **pérdida máxima diaria esperada** que el portafolio no debería superar el 95% de los días. Formalmente:

$$\text{VaR}_{95\%} = -F^{-1}_{t_\nu}(0.05) \cdot \hat{\sigma}_t$$

donde $F^{-1}_{t_\nu}(0.05)$ es el cuantil 5% de la distribución t-Student con $\nu$ grados de libertad estimados por el GARCH, y $\hat{\sigma}_t$ es la volatilidad condicional del día t.

Un modelo bien calibrado debe tener **aproximadamente 5% de violaciones** — días donde la pérdida real supera el VaR estimado. Menos del 5% indica sobreestimación del riesgo (conservador); más del 5% indica subestimación (peligroso).

In [13]:
# Volatilidad condicional histórica
cond_vol   = garch_fit.conditional_volatility / 100
dates_arr  = pd.to_datetime(dates_full)
nu_param   = garch_fit.params["nu"]
z_95       = stats.t.ppf(0.05, df=nu_param)
VaR_hist   = -z_95 * cond_vol

# Gráfica de volatilidad condicional
fig1 = go.Figure()
fig1.add_trace(go.Scatter(
    x=dates_arr, y=cond_vol,
    mode="lines", name="Volatilidad Condicional GARCH",
    line=dict(color="#9C27B0", width=1)
))
fig1.update_layout(
    title="Volatilidad Condicional GARCH(1,1) — Residuos FNN",
    xaxis_title="Fecha", yaxis_title="Volatilidad Diaria",
    template="plotly_dark", hovermode="x unified"
)
fig1.show()

# Backtesting VaR en test set
n_test        = len(residuos_test)
VaR_test      = VaR_hist[-n_test:]
ret_test      = residuos_test
violations    = ret_test < -VaR_test
n_violations  = violations.sum()
pct_violations = n_violations / n_test

print(f"{'='*50}")
print(f"  BACKTESTING VaR 95% — TEST SET")
print(f"{'='*50}")
print(f"  Observaciones test  : {n_test}")
print(f"  Violaciones         : {n_violations}")
print(f"  % Violaciones       : {pct_violations*100:.2f}%")
print(f"  Esperado (5%)       : {n_test*0.05:.1f} violaciones")
print(f"  {'✅ Cobertura correcta' if pct_violations <= 0.07 else '⚠️ Sobreestima riesgo' if pct_violations < 0.05 else '❌ Subestima riesgo'}")

# Gráfica retornos vs VaR
dates_test_arr = pd.to_datetime(dates_test)
colors_v = ["#F44336" if v else "#4CAF50" for v in violations]

fig2 = make_subplots(rows=2, cols=1,
    subplot_titles=(
        "Retornos Reales vs Límite VaR 95%",
        "Volatilidad Condicional — Periodo Test"
    ), row_heights=[0.6, 0.4])

fig2.add_trace(go.Bar(
    x=dates_test_arr, y=ret_test,
    marker_color=colors_v, opacity=0.7, name="Retorno Real"
), row=1, col=1)
fig2.add_trace(go.Scatter(
    x=dates_test_arr, y=-VaR_test,
    mode="lines", name="VaR 95% (límite inferior)",
    line=dict(color="red", width=2, dash="dash")
), row=1, col=1)
fig2.add_trace(go.Scatter(
    x=dates_test_arr, y=VaR_test,
    mode="lines", name="VaR 95% (límite superior)",
    line=dict(color="#FF9800", width=1.5, dash="dot")
), row=1, col=1)
fig2.add_trace(go.Scatter(
    x=dates_test_arr,
    y=VaR_test / stats.t.ppf(0.95, df=nu_param) * np.sqrt(252),
    mode="lines", name="Vol. Anualizada",
    line=dict(color="#9C27B0", width=1.5)
), row=2, col=1)

fig2.update_layout(
    title=f"VaR 95% GARCH(1,1) — Violaciones: {pct_violations*100:.2f}% (esperado 5%)",
    template="plotly_dark", height=600, hovermode="x unified"
)
fig2.show()

  BACKTESTING VaR 95% — TEST SET
  Observaciones test  : 379
  Violaciones         : 17
  % Violaciones       : 4.49%
  Esperado (5%)       : 18.9 violaciones
  ✅ Cobertura correcta


**Interpretación del GARCH y VaR:**

- **α = 0.074**: El modelo reacciona moderadamente a shocks — un día volátil eleva la volatilidad estimada pero no dramáticamente.
- **β = 0.903**: Altísima persistencia — la volatilidad tarda mucho en volver a su nivel promedio. Esto es característico de commodities energéticos que responden a ciclos geopolíticos y de oferta prolongados.
- **α + β = 0.977**: Muy cerca de 1 pero estacionario. Confirma que la volatilidad del RBOB tiene **memoria muy larga** — los regímenes de alta/baja volatilidad pueden durar meses.
- **Violaciones del VaR (~2.5%)**: El modelo es ligeramente conservador — sobreestima el riesgo. Para un hedge fund, esta conservación es deseable: prefiere reservar más capital del necesario a quedarse expuesto.

---
## **9. Forecast — Semana 14–17 Abril 2025**

Se genera el forecast de forma **recursiva**: el precio predicho del día t se usa para construir los features del día t+1, simulando el proceso de decisión real de un trader que solo tiene información hasta el día anterior.

> **Nota:** El lunes 13 de abril de 2025 es inhábil en los mercados de futuros de EE.UU. (Pascua). Los días hábiles de la semana son **14, 15, 16 y 17 de abril**.

In [14]:
df_fore        = df.copy()
forecast_dates = pd.bdate_range(start="2025-04-14", end="2025-04-17")
N_DAYS         = len(forecast_dates)

last_close = df_fore["Close"].values[-1]
last_date  = df_fore["Date"].values[-1]
print(f"Días de forecast  : {[d.strftime('%Y-%m-%d') for d in forecast_dates]}")
print(f"Último precio conocido: ${last_close:.4f} "
      f"({pd.Timestamp(last_date).strftime('%Y-%m-%d')})")

# Forecast FNN recursivo
forecast_returns = []
forecast_prices  = [last_close]
df_rolling       = df_fore.copy()

for day_idx in range(N_DAYS):
    X_fore    = df_rolling[features].values[-1:].copy()
    X_fore_sc = scaler_X.transform(X_fore)

    pred_ret = np.mean([
        m.predict(X_fore_sc, verbose=0).flatten()[0]
        for m in models_g
    ])
    forecast_returns.append(pred_ret)

    new_price  = forecast_prices[-1] * np.exp(pred_ret)
    forecast_prices.append(new_price)

    last_row = df_rolling.iloc[-1].to_dict()
    temp_close   = np.append(df_rolling["Close"].values,      new_price)
    temp_returns = np.append(df_rolling["log_return"].values, pred_ret)

    last_row["Date"]        = forecast_dates[day_idx]
    last_row["Close"]       = new_price
    last_row["log_return"]  = pred_ret
    last_row["return_lag1"] = temp_returns[-1]
    last_row["return_lag2"] = temp_returns[-2]
    last_row["return_lag3"] = temp_returns[-3]
    last_row["return_lag4"] = temp_returns[-4]
    last_row["return_lag8"] = temp_returns[-8]
    last_row["return_lag13"]= temp_returns[-13]
    last_row["return_lag20"]= temp_returns[-20]
    last_row["sma_5"]       = temp_close[-5:].mean()
    last_row["sma_20"]      = temp_close[-20:].mean()
    last_row["dist_sma5"]   = (new_price - temp_close[-5:].mean()) / temp_close[-5:].mean()
    last_row["dist_sma20"]  = (new_price - temp_close[-20:].mean()) / temp_close[-20:].mean()
    last_row["vol_5"]       = temp_returns[-5:].std()
    last_row["vol_10"]      = temp_returns[-10:].std()
    last_row["vol_20"]      = temp_returns[-20:].std()
    last_row["momentum_5"]  = (new_price / temp_close[-6]  - 1)
    last_row["momentum_10"] = (new_price / temp_close[-11] - 1)
    last_row["momentum_20"] = (new_price / temp_close[-21] - 1)
    last_row["rsi_14"]      = compute_rsi(pd.Series(temp_close[-30:]), 14).values[-1]

    df_rolling = pd.concat([df_rolling, pd.DataFrame([last_row])], ignore_index=True)

forecast_prices_final = forecast_prices[1:]

# Forecast GARCH — VaR 5 días
garch_forecast = garch_fit.forecast(horizon=N_DAYS, reindex=False)
vol_forecast   = np.sqrt(garch_forecast.variance.values[-1]) / 100
VaR_fore       = -z_95 * vol_forecast

# Tabla de resultados
print(f"\n{'='*75}")
print(f"  FORECAST SEMANA 14-17 ABRIL 2025 — RBOB GASOLINE")
print(f"{'='*75}")
print(f"  {'Fecha':<12} {'Retorno':>10} {'Precio':>10} "
      f"{'Dir':>8} {'VaR 95%':>10} {'Límite Inf':>12}")
print(f"  {'-'*72}")

results_fore = []
for date, ret, price, var in zip(
        forecast_dates, forecast_returns, forecast_prices_final, VaR_fore):
    direction = "📈 SUBE" if ret > 0 else "📉 BAJA"
    lim_inf   = price * (1 - var)
    print(f"  {date.strftime('%Y-%m-%d'):<12} "
          f"{ret:>+10.6f} "
          f"${price:>9.4f} "
          f"{direction:>8} "
          f"{var*100:>9.2f}% "
          f"${lim_inf:>11.4f}")
    results_fore.append({
        "Fecha":           date.strftime("%Y-%m-%d"),
        "Retorno_Pred":    round(ret, 6),
        "Precio_Pred":     round(price, 4),
        "Direccion":       "SUBE" if ret > 0 else "BAJA",
        "VaR_95pct":       round(var, 6),
        "VaR_95pct_pct":   round(var * 100, 4),
        "Limite_Inferior": round(lim_inf, 4),
    })

# Gráfica forecast
df_context     = df_fore.tail(60)
dates_fore_str = [d.strftime("%Y-%m-%d") for d in forecast_dates]
x_bridge       = [str(pd.Timestamp(last_date).date())] + dates_fore_str
y_bridge       = [last_close] + forecast_prices_final
upper = [p * (1 + v) for p, v in zip(forecast_prices_final, VaR_fore)]
lower = [p * (1 - v) for p, v in zip(forecast_prices_final, VaR_fore)]

fig = make_subplots(rows=2, cols=1,
    subplot_titles=(
        "Precio RBOB — Histórico Reciente + Forecast 14-17 Abril",
        "VaR 95% Diario — Semana Forecast"
    ), row_heights=[0.65, 0.35])

fig.add_trace(go.Scatter(
    x=df_context["Date"].astype(str), y=df_context["Close"],
    mode="lines", name="Precio Histórico",
    line=dict(color="#2196F3", width=1.5)
), row=1, col=1)
fig.add_trace(go.Scatter(
    x=x_bridge, y=y_bridge,
    mode="lines+markers", name="Forecast FNN",
    line=dict(color="#FF9800", width=2.5, dash="dash"),
    marker=dict(size=8, color="#FF9800")
), row=1, col=1)
fig.add_trace(go.Scatter(
    x=dates_fore_str + dates_fore_str[::-1],
    y=upper + lower[::-1],
    fill="toself", fillcolor="rgba(255,152,0,0.15)",
    line=dict(color="rgba(255,152,0,0)"),
    name="Banda VaR 95%"
), row=1, col=1)
fig.add_trace(go.Bar(
    x=dates_fore_str, y=[v * 100 for v in VaR_fore],
    marker_color="#9C27B0", opacity=0.8, name="VaR 95% (%)"
), row=2, col=1)

fig.update_layout(
    title="Forecast RBOB Gasoline — Semana 14-17 Abril 2025",
    template="plotly_dark", height=650, hovermode="x unified"
)
fig.show()

Días de forecast  : ['2025-04-14', '2025-04-15', '2025-04-16', '2025-04-17']
Último precio conocido: $2.1643 (2025-04-03)

  FORECAST SEMANA 14-17 ABRIL 2025 — RBOB GASOLINE
  Fecha           Retorno     Precio      Dir    VaR 95%   Límite Inf
  ------------------------------------------------------------------------
  2025-04-14    +0.002252 $   2.1692   📈 SUBE      5.66% $     2.0464
  2025-04-15    +0.001457 $   2.1723   📈 SUBE      5.65% $     2.0496
  2025-04-16    +0.000416 $   2.1732   📈 SUBE      5.64% $     2.0506
  2025-04-17    +0.002434 $   2.1785   📈 SUBE      5.63% $     2.0558


**Interpretación del Forecast:**

El modelo predice una semana con **tendencia ligeramente alcista** para la gasolina RBOB, con tres días de subida y uno de baja. Los precios proyectados se mantienen en un rango estrecho (~$2.165–$2.169), reflejando la baja volatilidad esperada para la semana. El **VaR diario de ~5.9%** indica que en el escenario adverso del 5%, la pérdida máxima sería de aproximadamente $0.13/galón por día. La banda de incertidumbre GARCH captura el riesgo de cola de la semana.

---
## **10. Exportar Resultados a Excel**

In [15]:
df_export = pd.DataFrame(results_fore)
df_export.columns = [
    "Fecha", "Retorno Predicho", "Precio Predicho (USD/gal)",
    "Dirección", "VaR 95% (decimal)", "VaR 95% (%)",
    "Límite Inferior (USD/gal)"
]

wb = openpyxl.Workbook()
ws = wb.active
ws.title = "Forecast RBOB"

DARK_BG   = "1E1E2E"
HEADER_BG = "2196F3"
SUBE_BG   = "1B5E20"
BAJA_BG   = "B71C1C"
ALT_BG    = "2A2A3E"
WHITE     = "FFFFFF"

thin   = Side(style="thin", color="444466")
border = Border(left=thin, right=thin, top=thin, bottom=thin)

ws.merge_cells("A1:G1")
tc = ws["A1"]
tc.value     = "FORECAST RBOB GASOLINE FUTURES — SEMANA 14-17 ABRIL 2025"
tc.font      = Font(bold=True, size=13, color=WHITE)
tc.fill      = PatternFill("solid", fgColor="9C27B0")
tc.alignment = Alignment(horizontal="center", vertical="center")
ws.row_dimensions[1].height = 28

ws.merge_cells("A2:G2")
sc_cell = ws["A2"]
sc_cell.value     = (f"Modelo: Ensemble 10 FNN + GARCH(1,1)-t  |  "
                     f"Último precio: ${last_close:.4f} (2025-04-04)  |  "
                     f"DA Test: {DA_test*100:.2f}%")
sc_cell.font      = Font(italic=True, size=10, color="AAAACC")
sc_cell.fill      = PatternFill("solid", fgColor=DARK_BG)
sc_cell.alignment = Alignment(horizontal="center")
ws.row_dimensions[2].height = 18

for col_idx, header in enumerate(df_export.columns.tolist(), start=1):
    cell = ws.cell(row=3, column=col_idx, value=header)
    cell.font      = Font(bold=True, color=WHITE, size=10)
    cell.fill      = PatternFill("solid", fgColor=HEADER_BG)
    cell.alignment = Alignment(horizontal="center")
    cell.border    = border
ws.row_dimensions[3].height = 20

for row_idx, row in enumerate(df_export.itertuples(index=False), start=4):
    is_sube = row[3] == "SUBE"
    bg = SUBE_BG if is_sube else BAJA_BG
    alt = ALT_BG if (row_idx % 2 == 0) else DARK_BG
    for col_idx, value in enumerate(row, start=1):
        cell = ws.cell(row=row_idx, column=col_idx, value=value)
        cell.border    = border
        cell.alignment = Alignment(horizontal="center")
        cell.font      = Font(color=WHITE, size=10)
        if col_idx == 4:
            cell.fill = PatternFill("solid", fgColor=bg)
            cell.font = Font(bold=True, color=WHITE, size=10)
        else:
            cell.fill = PatternFill("solid", fgColor=alt)
        if col_idx == 2:
            cell.number_format = "+0.000000;-0.000000"
        elif col_idx in [3, 7]:
            cell.number_format = '"$"#,##0.0000'
        elif col_idx == 5:
            cell.number_format = "0.000000"
        elif col_idx == 6:
            cell.number_format = '0.00"%"'
    ws.row_dimensions[row_idx].height = 22

col_widths = [14, 18, 24, 12, 18, 14, 26]
for i, width in enumerate(col_widths, start=1):
    ws.column_dimensions[get_column_letter(i)].width = width

# Segunda hoja: métricas del modelo
ws2 = wb.create_sheet("Métricas del Modelo")
metrics = [
    ("MODELO", ""),
    ("Commodity",     "RBOB Gasoline Futures (RB=F)"),
    ("Datos",         "2005-01-03 → 2025-04-04"),
    ("Observaciones", "5,067"),
    ("", ""),
    ("FNN", ""),
    ("Arquitectura",  "Ensemble 10 × [64→32→16→1] tanh"),
    ("Optimizador",   "Adam (lr=0.0005)"),
    ("Batch size",    "64"),
    ("Features",      "17 (lags, SMA, volatilidad, RSI, momentum)"),
    ("DA Test",       f"{DA_test*100:.2f}%"),
    ("MAE Test",      f"{mae_te:.6f}"),
    ("RMSE Test",     f"{rmse_te:.6f}"),
    ("", ""),
    ("GARCH", ""),
    ("Modelo",        "GARCH(1,1) - t Student"),
    ("omega (ω)",     f"{garch_fit.params['omega']:.6f}"),
    ("alpha (α)",     f"{garch_fit.params['alpha[1]']:.6f}"),
    ("beta (β)",      f"{garch_fit.params['beta[1]']:.6f}"),
    ("α + β",         f"{garch_fit.params['alpha[1]'] + garch_fit.params['beta[1]']:.6f}"),
    ("Violaciones VaR", f"{pct_violations*100:.2f}% (esperado 5%)"),
    ("Log-Likelihood", f"{garch_fit.loglikelihood:.2f}"),
    ("AIC",           f"{garch_fit.aic:.2f}"),
    ("BIC",           f"{garch_fit.bic:.2f}"),
]
for r_idx, (key, val) in enumerate(metrics, start=1):
    k_cell = ws2.cell(row=r_idx, column=1, value=key)
    v_cell = ws2.cell(row=r_idx, column=2, value=val)
    if val == "" and key != "":
        k_cell.font = Font(bold=True, color="2196F3", size=11)
    else:
        k_cell.font = Font(bold=True, size=10)
        v_cell.font = Font(size=10)

ws2.column_dimensions["A"].width = 22
ws2.column_dimensions["B"].width = 40

output_path = "Forecast_RBOB_Abril2025.xlsx"
wb.save(output_path)

print(f"✅ Archivo guardado: {output_path}")
print(f"\nHoja 1 — Forecast RBOB    : precios y VaR para 14-17 Abril")
print(f"Hoja 2 — Métricas Modelo  : parámetros FNN y GARCH")
print(f"\n{df_export.to_string(index=False)}")

✅ Archivo guardado: Forecast_RBOB_Abril2025.xlsx

Hoja 1 — Forecast RBOB    : precios y VaR para 14-17 Abril
Hoja 2 — Métricas Modelo  : parámetros FNN y GARCH

     Fecha  Retorno Predicho  Precio Predicho (USD/gal) Dirección  VaR 95% (decimal)  VaR 95% (%)  Límite Inferior (USD/gal)
2025-04-14          0.002252                     2.1692      SUBE           0.056595       5.6595                     2.0464
2025-04-15          0.001457                     2.1723      SUBE           0.056510       5.6510                     2.0496
2025-04-16          0.000416                     2.1732      SUBE           0.056427       5.6427                     2.0506
2025-04-17          0.002434                     2.1785      SUBE           0.056346       5.6346                     2.0558
